evap/staff/importers/enrollment.py

```diff
@@ -0,0 +1,764 @@
+from collections import defaultdict
+from dataclasses import dataclass, fields
+from typing import Dict, Iterable, List, Optional, Set, Union
+
+from django.conf import settings
+from django.db import transaction
+from django.utils.html import format_html, format_html_join
+from django.utils.translation import gettext as _
+from typing_extensions import TypeGuard
+
+from evap.evaluation.models import Contribution, Course, CourseType, Degree, Evaluation, Semester, UserProfile
+from evap.evaluation.tools import ilen
+from evap.staff.tools import create_user_list_html_string_for_message
+
+from .base import (
+    Checker,
+    ConvertExceptionsToMessages,
+    ExcelFileLocation,
+    ExcelFileRowMapper,
+    FirstLocationAndCountTracker,
+    ImporterLog,
+    ImporterLogEntry,
+    InputRow,
+    RowCheckerMixin,
+)
+from .user import (
+    UserData,
+    UserDataEmptyFieldsChecker,
+    UserDataMismatchChecker,
+    UserDataValidationChecker,
+    get_user_profile_objects,
+    update_existing_and_create_new_user_profiles,
+)
+
+
+class InvalidValue:
+    pass
+
+
+invalid_value = InvalidValue()
+
+
+@dataclass
+class CourseData:
+    """Holds information about a course, retrieved from an import file."""
+
+    name_de: str
+    name_en: str
+    degrees: Union[Set[Degree], InvalidValue]
+    course_type: Union[CourseType, InvalidValue]
+    is_graded: Union[bool, InvalidValue]
```

Comment:
This can be a bit nicer, using
```python
T = TypeVar("T")
MaybeInvalid = Union[T, InvalidValue]
```

and then `MaybeInvalid[CourseType]` and such